theano/tensor/nnet/tests/test_ctc.py

```diff
@@ -0,0 +1,117 @@
+from __future__ import (division, absolute_import, print_function)
+
+import unittest
+import numpy as np
+import numpy.random
+
+import theano
+import theano.tensor as T
+from theano.tests import unittest_tools as utt
+from theano.tensor.nnet.ctc import (ctc_enabled, ctc)
+
+
+class TestCTC(unittest.TestCase):
+    """
+    Test Baidu CTC wrapper implementation.
+
+    Expected values for costs and gradients are obtained through an external
+    C implementation, that uses the library directly.
+    """
+
+    def setUp(self):
+        if not ctc_enabled:
+            self.skipTest('Optional library warp-ctc not available')
+
+    def run_ctc(self, activations, labels, input_length, expected_costs, expected_grads):
+        # Check if softmax probabilites are approximately equal to the gradients
+        # of the activations, using utt.assert_allclose(a, b)
+
+        # Create symbolic variables
+        t_activations = theano.shared(activations, name="activations")
+        t_activation_times = theano.shared(input_length, name="activation_times")
+        t_labels = theano.shared(labels, name="labels")
+
+        t_cost = ctc(t_activations, t_labels, t_activation_times)
+        # Symbolic gradient of CTC cost
+        t_grad = T.grad(T.mean(t_cost), t_activations)
+        # Compile symbolic functions
+        train = theano.function([], [t_cost, t_grad])
+
+        cost, grad = train()
+
+        utt.assert_allclose(expected_grads, grad)
```

Comment:
The grad is computed for the mean of costs, not for costs themselves. Currently, the expected gradients you give is the gradient for the cost.

To get the expected gradient for the mean, I think we could just change this line:
```python
utt.assert_allclose(expected_grads / input_length.shape[0], grad)
```
(gradient of (mean of (costs)) is mean of (gradients of (costs)), and there are `input_length.shape[0]` costs, I think)